core/container.go

```diff
@@ -159,6 +170,43 @@ func (container *Container) WithMountedFile(ctx context.Context, target string,
 	return container.withMounted(ctx, target, source)
 }
 
+func (container *Container) WithMountedCache(ctx context.Context, target string, source *Directory) (*Container, error) {
+	payload, err := container.ID.decode()
+	if err != nil {
+		return nil, err
+	}
+
+	mount := ContainerMount{
+		Target:           target,
+		CacheID:          fmt.Sprintf("%s:%s", container.ID, target),
```

Comment:
I agree that we don't want to ship this as-is, but I don't want to figure out that API as part of this PR as there's already quite a lot tied into it, and some of it unblocks later things (I have a `secrets` branch based on this one). I could just remove this implementation, I only kept it because it's one of the earlier commits I made.

For `CacheVolume` are we thinking something like this?

```graphql
scalar CacheID

extend type Query {
  cache(id: CacheID!): CacheVolume! # for consistency

  # for generating an ID (better word than tokens welcome)
  cacheFromTokens(tokens: [String!]!): CacheVolume!
}

type CacheVolume {
  id: CacheID!

  # maybe more fieldslater?
}

extend type Container {
  withMountedCache(path: String!, id: CacheID!, source: DirectoryID): Container!
}
```

Not opinionated about names/etc., just trying to get an idea of the general shape.

Reply:
Yeah I think that's basically what I was imagining in terms of API.

I don't think it's worth creating rebasing nightmares by removing this commit as long as we get the newer api in before preview, which seems pretty doable? If we don't though then we should just hide this API from users probably. Either way, doesn't have to block this PR imo

Reply:
Created https://github.com/dagger/dagger/issues/3270 for this